src/diffusers/utils/inputs.py

```diff
@@ -0,0 +1,137 @@
+# Copyright 2024 The HuggingFace Team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Generic utilities
+"""
+
+from collections import OrderedDict
+from dataclasses import fields, is_dataclass
+from typing import Any, Tuple
+
+import numpy as np
+
+from .import_utils import is_torch_available, is_torch_version
+
+
+def is_tensor(x) -> bool:
+    """
+    Tests if `x` is a `torch.Tensor` or `np.ndarray`.
+    """
+    if is_torch_available():
+        import torch
+
+        if isinstance(x, torch.Tensor):
+            return True
+
+    return isinstance(x, np.ndarray)
+
+
+class BaseInput(OrderedDict):
```

Comment:
nice!
is this same (very similar) to BaseOutput?

Reply:
Yes, there's minimal change, the only difference is that it keeps None values.

```diff
-class BaseInput(OrderedDict):
+class BaseOutput(OrderedDict):
     """
     Base class for all model outputs as dataclass. Has a `__getitem__` that allows indexing by integer or slice (like a
-    tuple) or strings (like a dictionary). Otherwise behaves like a regular Python dictionary.
-
-    Unlike BaseOutput, BaseInput will **not** ignore the `None` attributes.
+    tuple) or strings (like a dictionary) that will ignore the `None` attributes. Otherwise behaves like a regular
+    Python dictionary.
 
     <Tip warning={true}>
 
-    You can't unpack a [`BaseInput`] directly. Use the [`~utils.BaseInput.to_tuple`] method to convert it to a tuple
+    You can't unpack a [`BaseOutput`] directly. Use the [`~utils.BaseOutput.to_tuple`] method to convert it to a tuple
     first.
 
     </Tip>
@@ -90,7 +89,8 @@ class BaseInput(OrderedDict):
         else:
             for field in class_fields:
                 v = getattr(self, field.name)
-                self[field.name] = v
+                if v is not None:
+                    self[field.name] = v
 
     def __delitem__(self, *args, **kwargs):
         raise Exception(f"You cannot use ``__delitem__`` on a {self.__class__.__name__} instance.")
@@ -112,7 +112,7 @@ class BaseInput(OrderedDict):
             return self.to_tuple()[k]
 
     def __setattr__(self, name: Any, value: Any) -> None:
-        if name in self.keys():
+        if name in self.keys() and value is not None:
             # Don't call self.__setitem__ to avoid recursion errors
             super().__setitem__(name, value)
         super().__setattr__(name, value)
@@ -132,6 +132,6 @@ class BaseInput(OrderedDict):
 
     def to_tuple(self) -> Tuple[Any, ...]:
         """
-        Convert self to a tuple containing all the attributes/keys.
+        Convert self to a tuple containing all the attributes/keys that are not `None`.
         """
         return tuple(self[k] for k in self.keys())
```

```python
from dataclasses import dataclass
from diffusers.image_processor import PipelineImageInput
from diffusers.utils import BaseOutput, BaseInput

@dataclass
class ControlNetUnionBaseInput(BaseInput):
    """
    The image input of [`ControlNetUnionModel`]:

    - 0: openpose
    - 1: depth
    - 2: hed/pidi/scribble/ted
    - 3: canny/lineart/anime_lineart/mlsd
    - 4: normal
    - 5: segment
    """

    openpose: PipelineImageInput = None
    depth: PipelineImageInput = None
    hed: PipelineImageInput = None
    canny: PipelineImageInput = None
    normal: PipelineImageInput = None
    segment: PipelineImageInput = None


@dataclass
class ControlNetUnionBaseOutput(BaseOutput):
    """
    The image input of [`ControlNetUnionModel`]:

    - 0: openpose
    - 1: depth
    - 2: hed/pidi/scribble/ted
    - 3: canny/lineart/anime_lineart/mlsd
    - 4: normal
    - 5: segment
    """

    openpose: PipelineImageInput = None
    depth: PipelineImageInput = None
    hed: PipelineImageInput = None
    canny: PipelineImageInput = None
    normal: PipelineImageInput = None
    segment: PipelineImageInput = None

union_input = ControlNetUnionBaseInput()
union_output = ControlNetUnionBaseOutput()
```
```python
>>> union_output[0]
IndexError: tuple index out of range
>>> union_output.to_tuple()
()
>>> union_output.keys()
odict_keys([])
>>> union_input[0]
>>> union_input.to_tuple()
(None, None, None, None, None, None)
>>> odict_keys(['openpose', 'depth', 'hed', 'canny', 'normal', 'segment'])
```

Reply:
do you think we can just do this - I'm not sure this is something applicable for other models so maybe don't have to make a `BaseInput`, no?

```python
@dataclass
class ControlNetUnionInput:
    """
    The image input of [`ControlNetUnionModel`]:

    - 0: openpose
    - 1: depth
    - 2: hed/pidi/scribble/ted
    - 3: canny/lineart/anime_lineart/mlsd
    - 4: normal
    - 5: segment
    """
    openpose: Optional[PipelineImageInput] = None
    depth: Optional[PipelineImageInput] = None
    hed: Optional[PipelineImageInput] = None
    canny: Optional[PipelineImageInput] = None
    normal: Optional[PipelineImageInput] = None
    segment: Optional[PipelineImageInput] = None

    def __len__(self) -> int:
        return len(vars(self))

    def __iter__(self):
        return iter(vars(self))

    def __getitem__(self, key):
        return getattr(self, key)
```